### PR TITLE
Do not return nonexistent kinds in `SyntaxFacts`' methods

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Fixes: https://github.com/dotnet/roslyn/issues/72300

### CODE REVIEW

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs
Comment (DoctorKrolic) [inline src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs:3607]: I didn't find any specific test file for `SyntaxFacts` class, so I put it here (e.g. test above uses `SyntaxFacts`)

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:1151]: Maybe add `Debug.Assert(((SyntaxKind)i).ToString() != i.ToString())`

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:1156]: Simplified your suggestion to `Debug.Assert(Enum.IsDefined(typeof(SyntaxKind), (SyntaxKind)i))` (had to use version with `typeof` due to the fact, that generic overload is not present in `netstandard2.0`). Does the same things, but reads a lot easier. Also since there are multiple methods, which loops through numbers like this added such guard to all of them. Not sure, whether this will be classified as an "unrelated" change by the compiler team, but since we faced bug of this nature, I guess, it's better to add such guard rails everywhere

File: src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Syntax/LexicalAndXml/LexicalTests.cs]: Actually, found out that there is `SyntaxTests` for that. Moving the test

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:1171]: Adding additional asserts revealed that are a lot unexisting kinds are returned from `GetPunctuationKinds()` method

**Comment by AlekseyTs** [comment]
> Do not return unexisting kinds in SyntaxFacts' methods

Could you please elaborate what the problem is and open an issue that describes a user-facing problem? Even though there is no member in the enum, the number is still reserved. And

**Comment by DoctorKrolic** [comment]
> Could you please elaborate what the problem is

I was working on a fix for https://github.com/dotnet/roslyn/issues/72258. I decided to populate all C# keywords' texts upfront, so I can later return them as completions. In order to achieve that I used `SyntaxFacts.GetKeywordKinds()` method and then saved keyword text by calling `SyntaxFacts.GetText(kind)` on each kind. When it came to verifying my work I noticed that there was 1 completion with empty text. A quick debug session showed that `GetKeywordKinds` has unexisting kind `8441` as part of returned collection of kinds. This led to this PR being opened. Then Youssef suggested adding guarding assert and I added such for all methods with similar integer looping logic. Failed CI then showed that `SyntaxFacts.GetPunctuationKinds()` also returns unexisting kinds and a lot of them. After fixing it I thought that corresponding `SyntaxFacts.IsXXX` checks should also not assume anything about unexisting kinds. `SyntaxFacts.IsKeywordKind()` already specifies all contextual keywords 1 by 1, so it didn't need fixing. `SyntaxFacts.IsPunctuation()` on the other hand did a range check, so I modified it to exclude unexisting kinds there. Now we are at this state.

> open an issue that describes a user-facing problem

> Even though there is no member in the enum, the number is still reserved

It is perfectly fine to reserve number ranges for different kinds of enum values. I would say, it is also fine to return unexisting kinds in an internal API, since this fact can then be shared as a common knowledge in the team, so additional checks can be made on caller side if necessary. However, this is public API and I can say with a very high level of certainty that consumers of public API by default assume that if API returns an enum value, this is an actual existing value of that enum. So I would classify this as a bug regardless of the user scenario it was discovered in (and I provided the way I personally discovered it)

**Comment by AlekseyTs** [comment]
> So I would classify this as a bug regardless of the user scenario it was discovered in (and I provided the way I personally discovered it)

I am afraid we still need an issue, especially as you classify the behavior as a bug.

**Comment by DoctorKrolic** [comment]
> I am afraid we still need an issue, especially as you classify the behavior as a bug.

Filled https://github.com/dotnet/roslyn/issues/72300 and https://github.com/dotnet/roslyn/issues/72301. I'll put this PR into a draft state until both issues are discussed and approved as actual bugs. Personally I feel _very_ strongly about the first one being a bug and less strongly about the second one (but it is still not normal behavior IMO)

**Comment by DoctorKrolic** [comment]
@AlekseyTs https://github.com/dotnet/roslyn/issues/72300 has been classified as a bug. I've reverted other changes in this PR, so it only fixes that issue. This is now ready for review

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:179]: >DotDotToken [](http://example.com/codeflow?start=74&length=11)

I think that in the SyntaxKind enum, around declarations of kinds that are hardcoded in implementation of this method, we should add comments describing what changes to the enum should lead to changes in this function as well.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 6)

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:283]: > What way of testing do you think is better here?

I am leaning towards the first way, but the range should be defined differently by comparison to ```GetPunctuationKinds```. The test will have limited value if it is based on the same logic. Perhaps, based on the kind immediately preceding the first punctuation and the kind immediately following the last punctuation.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:311]: >ToArray [](http://example.com/codeflow?start=87&length=7)

A HashSet of integer values (not enum values) is probably be more efficient for the purpose of ```Contains``` checks.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:314]: >Assert.DoesNotContain((SyntaxKind)((int)firstExpectedKind - 1), kindsFromMethod); [](http://example.com/codeflow?start=12&length=81)

It is not clear why we are checking only against two values outside of the range. What prevents us from iterating through the result and asserting all values are within the expected range?

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:299]: >SyntaxKind.FileKeyword [](http://example.com/codeflow?start=92&length=22)

It feels like the test will require less maintenance if the end range value is exclusive, i.e. ```ElifKeyword``` for this range.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:300]: >// This case doesn't cover all possible preprocessor kinds [](http://example.com/codeflow?start=121&length=58)

Then, perhaps, it should have its own dedicated test that provides proper coverage?

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:298]: This thread was marked as resolved, but I didn't see any response in the code or otherwise. Reactivating.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:311]: >Assert.InRange(kind, firstExpectedKind, lastExpectedKind); [](http://example.com/codeflow?start=16&length=58)

I am not sure what is going on with this test in commit 9. Instead of making incremental improvements to the test, it appears that the test got changed in a way that it no longer serves its requested purpose - "We should try to come up with a test that would fail when ```GetPunctuationKinds``` fails to return all expected kinds." What am I missing.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:321]: >SyntaxKind.RestoreKeyword [](http://example.com/codeflow?start=62&length=25)

Same suggestion as for the previous test in terms of future maintenance.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:316]: >TestGetPreprocessorKeywordKindsReturnsExpectedResults [](http://example.com/codeflow?start=20&length=53)

It looks like this test is also failing to fulfil the requested aspect - "a test that would fail when ```GetPreprocessorKeywordKinds``` fails to return all expected kinds"

**Comment by AlekseyTs** [comment]
Done with review pass (commit 9)

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs]: > It is not clear why we are checking only against two values outside of the range. What prevents us from iterating through the result and asserting all values are within the expected range?

I am not sure what has happened with this suggestion. It looks like instead of strengthening the check. I.e. making sure the returned kinds all are withing the expected range, the following checks were completely removed:
```
            // Check that the value right before the expected lower bound isn't contained in the enumerable returned from the target method
            Assert.DoesNotContain((SyntaxKind)((int)firstExpectedKind - 1), kindsFromMethod);
             // Check that the value right after the expected upper bound isn't contained in the enumerable returned from the target method
            Assert.DoesNotContain((SyntaxKind)((int)lastExpectedKind + 1), kindsFromMethod);
```

Could you, please, elaborate?

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:326]: >TestGetPreprocessorKeywordKindsReturnsExpectedResults [](http://example.com/codeflow?start=20&length=53)

Please strengthen the test by making sure the returned kinds all are withing the expected set.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 10)

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs]: I initialy understood this review as "we can replace current logic with simple in-range check". After doing that you said that the tests are no longer representative since they won't fail if the method fails to return all expected kinds. At the point of partially reverting my changes I just forgot that this was a thing.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs:329]: Done by removing expected members from the set and at the end asserting, that the set is empty

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxTests.cs]: Done by removing expected members from the set and at the end asserting, that the set is empty. It does the same thing as "in range" check since if something isn't in range it is gonna be leftover in the set and thus the assertion would fail

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs
Comment (DoctorKrolic) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:149]: The `SyntaxKind.ElifKeyword` - `SyntaxKind.RestoreKeyword` already contains `SyntaxKind.HiddenKeyword`, so I removed the duplicate. Found while working on tests

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 11), assuming CI is passing.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:22]: >(SyntaxKind)i [](http://example.com/codeflow?start=64&length=13)

nit: do we need the cast here?

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:22]: As far as I remember, there is an exception if we don't case, let me double check

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:22]: Yes, we do because the underlying type of `SyntaxKind` is `ushort`:

> System.ArgumentException : Enum underlying type and the object must be same type or object must be a String. Type passed in was 'System.Int32'; the enum underlying type was 'System.UInt16'.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 11)

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:22]: Thanks

**Comment by jcouv** [comment]
Merged/squashed. Thanks for the contribution!

**Pull request merged**
